pkg/generators/rules/lifecycle_tag.go

```diff
@@ -0,0 +1,242 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rules
+
+import (
+	"go/ast"
+	"go/parser"
+	"go/token"
+	gotypes "go/types"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"k8s.io/gengo/examples/set-gen/sets"
+	"k8s.io/gengo/types"
+)
+
+const (
+	// TagLifecycle is the comment tag prefix for specifying
+	// the API lifecycle info. Example tag format:
+	// +lifecycle:component=<component-name>,minVersion=<version>,status=<prerelease-status>,featureGate=<featuregate-name>
+	TagLifecycle = "lifecycle:component"
+
+	TagComponent   = "component"
+	TagMinVersion  = "minVersion"
+	TagStatus      = "status"
+	TagFeatureGate = "featureGate"
+
+	kubernetesComponentName = "kubernetes"
+
+	// as defined in https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go
+	kubernetesFeatureGateType = "featuregate.Feature"
+)
+
+var (
+	AllowedTagKeys = sets.NewString(TagComponent, TagMinVersion, TagStatus, TagFeatureGate)
+
+	// the status names must match the prereleases defined in
+	// k8s.io/component-base/featuregate
+	AllowedStatusNames = sets.NewString("alpha", "beta", "deprecated")
+
+	// regex for a valid k8s version
+	minVersionRegex = regexp.MustCompile(`^v[1-9][0-9]*\.(0|[1-9][0-9]*)$`)
+)
+
+// APILifecyleTag implements APIRule interface.
+type APILifecyleTag struct {
+	// FeatureGateFileNames is the list of files defining feature gates.
+	FeatureGateMap map[string]FeatureGateInfo
+}
+
+// FeatureGateInfo stores info about a particular feature gate.
+type FeatureGateInfo struct {
+	// status is the highest status for the feature gate.
+	status string
+	// minVersion is the minimum k8s version for the status.
+	minVersion string
+}
+
+// Name returns the name of APIRule
+func (l *APILifecyleTag) Name() string {
+	return "incorrect_lifecycle_tag"
+}
+
+// Validate evaluates API rule on type t and returns a list of field names in
+// the type that violate the rule. Empty field name [""] implies the entire
+// type violates the rule.
+func (l *APILifecyleTag) Validate(t *types.Type) ([]string, error) {
+	fields := make([]string, 0)
+
+	switch t.Kind {
+	case types.Struct:
+		for _, m := range t.Members {
+			commentTag := types.ExtractCommentTags("+", m.CommentLines)[TagLifecycle]
+			if commentTag != nil {
```

Comment:
for GA fields (`commentTag == nil`), report them as violation and add them to the exception file

Reply:
> for GA fields (commentTag == nil), report them as violation and add them to the exception file

I'm not sure I follow :thinking: 

From the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2558-publish-version-openapi/README.md#proposal), we don't publish this extension for GA fields so GA fields should not have this comment tag i.e. `commentTag == nil` would be expected and would not be violation.

> When an API field moves to GA, the feature gate and the comment tag for the API field are removed. This will also remove the x-kubernetes-api-lifecycle vendor extension for the field in OpenAPI.

Reply:
I thought https://github.com/kubernetes/enhancements/pull/2560#discussion_r608867989 was added to the KEP.

I expect the linter to report an error when a field doesn't have the tag. Otherwise how will we force API authors to add the tag on new fields?

Reply:
I realized that this was added before but got incorrectly removed when we decided not to support the extension for `Kind`s. :woman_facepalming: 

Added it back in https://github.com/kubernetes/enhancements/pull/2752. PTAL.

Also updated code in this PR to add GA fields to exceptions. This will lead to almost 4k additional lines in the violations file though - [example diff](https://gist.github.com/nikhita/ce4a36f40f8d166d695fbddd11df27ca). Not sure if that would be ok... :thinking:

Reply:
Thanks! I think that's okay. We need those many lines anyway as long as we want to distinguish GA fields and new Alpha/Beta fields. Maybe we could improve by putting them in a separate file, but I think currently a single file 1) serves the purpose of preventing people from adding violations with no good reason, 2) still allows people to navigate the file and fix violations for other rules if they want.